deltacat/compute/compactor/steps/materialize.py

```diff
@@ -164,49 +172,66 @@ def _materialize(
         mask = pa.array(mask_pylist)
         pa_table = pa_table.filter(mask)
 
-        # appending, sorting, taking, and dropping has 2-3X latency of a
-        # single filter on average, and thus provides better average
-        # performance than repeatedly filtering the table in dedupe task index
-        # order
-        dedupe_task_indices = chain.from_iterable(dedupe_task_idx_iter_tpl)
-        pa_table = sc.append_dedupe_task_idx_col(
-            pa_table,
-            dedupe_task_indices,
-        )
-        pa_sort_keys = [(sc._DEDUPE_TASK_IDX_COLUMN_NAME, "ascending")]
-        pa_table = pa_table.take(
-            pc.sort_indices(pa_table, sort_keys=pa_sort_keys),
-        )
-        pa_table = pa_table.drop(
-            [sc._DEDUPE_TASK_IDX_COLUMN_NAME]
-        )
+        if remainder_tables:
+            assert remainder_record_count < max_records_per_output_file, \
+                f"Total records in previous remainder should not exceed {max_records_per_output_file}."
+
+            # Skip materializing tables until we have reached 'max_records_per_output_file'
+            if remainder_record_count + len(pa_table) < max_records_per_output_file:
+                remainder_tables.append(pa_table)
+                remainder_record_count += len(pa_table)
+                continue
+
+            # 'max_records_per_output_file' will be exceeded with the
+            #  current file's records + previous remainder records.
+            records_to_fit = max_records_per_output_file - remainder_record_count
+            fitted_table = pa_table.slice(length=records_to_fit)
+            remainder_tables.append(fitted_table)
 
-        # Write manifests up to max_records_per_output_file
-        # TODO(raghumdani): Write exactly the same number of records into each file to
-        # produce a read-optimized view of the tables.
-        if compacted_tables and \
-                total_record_count + record_count > max_records_per_output_file:
-            materialized_results.append(_materialize(compacted_tables, total_record_count))
+            # Chop off records that are already written to disk
+            pa_table = pa_table.slice(offset=records_to_fit)
+
+        record_count = len(pa_table)
+        assert record_count > 0, f"Expected non-empty compacted PyArrow table"
```

Comment:
Would this always be true? What happens when `remainder_record_count + len(pa_table)` is equal to `max_records_per_output_file` ?

Reply:
No, thanks for catching (this is why we need unit tests) 🙂